### DISCUSSION TITLE
Support Response Interception

### DISCUSSION THREAD

**Discussion by barefootdeveloper** [mockoon/mockoon#477] [Ideas] [answered]
## Purpose
To modify the request response on the fly as the user makes requests. We can compare this to the intercept capability of Fiddler or [Burp Suite](https://portswigger.net/burp/documentation/desktop/tools/proxy/getting-started) except we will only focus on modifying the response.
## Proposal
Implement an environment level setting that enables or disables intercepts. It should support 3 settings:
  1. Disabled: Will not intercept any responses.
  2. Intercept all responses: Will intercept every response from the mock server.
  3. Intercept enabled routes: Will intercept select routes from the mock server.

Implement a route level setting to enable intercepts for a specific route (if interception is enabled).

Change the history screen to allow for controls that allow you to modify the response sent from the mock server.

![Snag_13d6ed91](https://user-images.githubusercontent.com/3580448/118414132-d8832d80-b670-11eb-8a98-326b32428c27.png)

The Response Body Editor should be expanded to allow the modification of other response properties, such as status and content-type. For convenience, content-length should be auto-calculated.

**Comment by 255kb** [comment, marked as answer]
This is a nice idea I've never used such feature. 

Two things on my mind: 
- I would rather put this feature on the route page as it is highly related to the other one influencing how responses are served: random and sequential responses. 
  ![image](https://user-images.githubusercontent.com/7489814/121156812-54742e00-c849-11eb-9fe5-7a100c2d78f5.png)
  Also, I find it a bit confusing that one is able to interact with the route responses in two different places (route page and environment logs), and I tend to consider the logs as a "static" page where you mainly go for debugging.
  So, I would try to add something there. Like this, we would also be able to reuse the body editor instead of adding somewhere else a way to edit the response.

- This will require some investigation on how to pause Expressjs request which is happening in the Main process (Node.js environment), until a user interaction is detected in the Renderer process, while still allowing subsequent requests to succeed. I think it will be quite a challenge!
